pkg/common/glob/glob.go

```diff
@@ -0,0 +1,123 @@
+package common
+
+import (
+	"fmt"
+
+	"github.com/gobwas/glob"
+)
+
+// Filter is a generic filter for excluding and including globs (limited
+// regular expressions). Exclusion takes precedence if both include and exclude
+// lists are provided.
+type Filter struct {
+	exclude []glob.Glob
+	include []glob.Glob
+}
+
+type globFilterOpt func(*Filter) error
+
+// WithExcludeGlobs adds exclude globs to the filter.
+func WithExcludeGlobs(excludes ...string) globFilterOpt {
+	return func(f *Filter) error {
+		for _, exclude := range excludes {
+			g, err := glob.Compile(exclude)
+			if err != nil {
+				return err
+			}
+			f.exclude = append(f.exclude, g)
+		}
+		return nil
+	}
+}
+
+// WithIncludeGlobs adds include globs to the filter.
+func WithIncludeGlobs(includes ...string) globFilterOpt {
+	return func(f *Filter) error {
+		for _, include := range includes {
+			g, err := glob.Compile(include)
+			if err != nil {
+				return err
+			}
+			f.include = append(f.include, g)
+		}
+		return nil
+	}
+}
+
+// NewGlobFilter creates a new Filter with the provided options.
+func NewGlobFilter(opts ...globFilterOpt) (*Filter, error) {
+	filter := &Filter{}
+	for _, opt := range opts {
+		if err := opt(filter); err != nil {
+			return nil, err
+		}
+	}
+	return filter, nil
+}
+
+// Pass returns whether the object is not in the include list or in the exclude
+// list.
+func (f *Filter) Pass(object string) bool {
+	if f == nil {
+		return false
+	}
+	exclude, include := len(f.exclude), len(f.include)
+	if exclude == 0 && include == 0 {
+		return false
+	} else if exclude > 0 && include == 0 {
+		return f.passExclude(object)
+	} else if exclude == 0 && include > 0 {
+		return f.passInclude(object)
+	} else {
+		if pass, err := f.passExcludeInclude(object); err == nil {
+			return pass
+		}
+		// Ambiguous case. Should we skip? Should we not skip?
+		// Let's skip since the user indicated they want *some* form of
+		// filtering.
+		// TODO: Maybe this could be configurable.
+		return true
+	}
+}
+
+// passExclude checks for explicitly excluded paths. This should only be called
+// when the include list is empty.
```

Comment:
implementing all this via four very-similar-yet-occasionally-inverted blocks, some of which have sneaky preconditions, dispatched via a four-clause if statement, is triggering my spidey sense. does the complexity come from differentiating between glob sets that are empty and glob sets that are non-empty but reject the candidate object?

Reply:
Yeah, it made the most sense in my head to differentiate the use cases.

* If I provide only an include list, I expect only matches to be scanned and everything else to be skipped.
* Likewise, if I only provide an exclude list, I expect the opposite behavior.
* If I provide both.. I'm not sure what to expect? Maybe an example would be overlapping globs. My expectation for the below is to scan everything under `/home/miccah/` except the tmp dir.
  * Include `/home/miccah/**`
  * Exclude `/home/miccah/tmp/**`

To maybe appease your spidey senses, the only exposed method is `Pass`. The other functions are not accessible outside the package and cannot be called.

Reply:
yeah, my brain is just trying to find a more compact mathematical representation of the logic. i _think_ what i want is this, which isn't actually possible and wouldn't perform optimally if it were:
1. adjust empty sets
  * if only the include set is empty, add a universal acceptance glob to it
  * if only the exclude set is empty, add a universal rejection glob to it <- this is the impossible part
2. test the candidate against both sets
  * if the exclude set accepts, reject the candidate
  * otherwise, iff the include set accepts, accept the candidate

this makes step (2) simpler, but at the cost of running unnecessary globs, and needing to use an undefinable "reject everything" glob. since what i want is impossible and slow, what you've got is probably best :)